code/controllers/subsystem/mapping.dm

```diff
@@ -0,0 +1,801 @@
+	var/sector_y = ceil(y / SECTOR_SIZE)
+	for(var/datum/allocation/A in allocations)
+		if(A.sector[1] == sector_x && A.sector[2] == sector_y && A.z == z)
+			return A
+	return z //return the z level if no allocation found
+
+/**
+ * Finds a suitable landing zone for a shuttle on a planet
+ *
+ * Searches the planet's sector for a flat area large enough to accommodate
+ * the specified dimensions, staying at least 11 tiles from sector edges.
+ * Returns a random valid location to provide variety.
+ *
+ * Arguments:
+ * * alloc - The planet allocation to search within
+ * * size - List containing [width, height] of the landing area needed
+ *
+ * Returns:
+ * * The top-left turf of a suitable landing zone, or null if no valid location found
+ */
+/datum/subsystem/mapping/proc/get_landing_zone(var/datum/allocation/alloc,var/list/size)
+	if (!alloc || !size || size.len != 2)
+		return null
+	var/x_dim = size[1]
+	var/y_dim = size[2]
+	var/list/turf/search_turfs = turfs_from_sector(alloc.sector, alloc.z)
+
+	// Get sector boundaries to calculate relative positions
+	var/list/bounds = get_sector_bounds(alloc.sector)
+	var/x_min = bounds["x_min"]
+	var/y_min = bounds["y_min"]
+
+	// Create matrix with relative coordinates
+	var/datum/turf_matrix[SECTOR_SIZE][SECTOR_SIZE]
+	for (var/turf/T in search_turfs)
+		var/rel_x = T.x - x_min + 1
+		var/rel_y = T.y - y_min + 1
+		turf_matrix[rel_x][rel_y] = T
+
+	// Define safe zone boundaries (accounting for edge buffer and shuttle size)
+	var/safe_x_min = LANDING_ZONE_EDGE_BUFFER + 1
+	var/safe_x_max = SECTOR_SIZE - LANDING_ZONE_EDGE_BUFFER - x_dim
+	var/safe_y_min = LANDING_ZONE_EDGE_BUFFER + 1
+	var/safe_y_max = SECTOR_SIZE - LANDING_ZONE_EDGE_BUFFER - y_dim
+
+	if(safe_x_max < safe_x_min || safe_y_max < safe_y_min)
+		return null // Not enough space for safe landing
+
+	// Create randomized search list within safe boundaries
+	var/list/search_positions = list()
+	for(var/rel_x = safe_x_min; rel_x <= safe_x_max; rel_x++)
+		for(var/rel_y = safe_y_min; rel_y <= safe_y_max; rel_y++)
+			var/turf/T = turf_matrix[rel_x][rel_y]
+			if(T && !iswall(T) && !istype(T, /turf/unsimulated/mineral))
+				search_positions += T
+
+	// Shuffle the search positions for randomization
+	if(!search_positions.len)
+		return null
+
+	search_positions = shuffle(search_positions)
+
+	// Search through randomized positions
+	for(var/turf/T in search_positions)
+		var/rel_x = T.x - x_min + 1
+		var/rel_y = T.y - y_min + 1
+		var/found = TRUE
+
+		for (var/dx = 0; dx < x_dim && found; dx++)
+			for (var/dy = 0; dy < y_dim && found; dy++)
+				var/check_x = rel_x + dx
+				var/check_y = rel_y + dy
+				if(check_x > SECTOR_SIZE || check_y > SECTOR_SIZE) // Out of sector bounds
+					found = FALSE
+					continue
+				var/turf/target = turf_matrix[check_x][check_y]
+				if (!target || !istype(target,T.type))
+					found = FALSE
+
+		if (found)
+			return T  // Return top-left turf of matching rectangle
+
+	return null
```

Comment:
That sounds like it should throw an error or something? Or some way to get feedback at the user-level? At the very least it should distinguish between "your shuttle is too fat to land here" and "there is no space at all"